Sources/NIO/System.swift

```diff
@@ -625,9 +625,21 @@ internal enum KQueue {
     @inline(never)
     @discardableResult
     public static func kevent(kq: CInt, changelist: UnsafePointer<kevent>?, nchanges: CInt, eventlist: UnsafeMutablePointer<kevent>?, nevents: CInt, timeout: UnsafePointer<Darwin.timespec>?) throws -> CInt {
-        return try syscall(blocking: false) {
-            sysKevent(kq, changelist, nchanges, eventlist, nevents, timeout)
-        }.result
+        do {
+            return try syscall(blocking: false) {
+                sysKevent(kq, changelist, nchanges, eventlist, nevents, timeout)
+            }.result
+        } catch let err as IOError where err.errnoCode == EINVAL {
+            // If kevent returned EINVAL, check if the timeout exceeded bounds impose by the kernel and if so throw
+            // a more meaningful error. These bounds are enforced by the Darwin kernel interval timer.
+            if let ts = timeout{
+                if ts.pointee.tv_sec < 0  || ts.pointee.tv_sec > 100_000_000 ||
+                     ts.pointee.tv_nsec < 0 || ts.pointee.tv_nsec >= 1_000_000_000 {
+                    throw IOError(errnoCode: err.errnoCode, reason: "kevent timeout exceeded bounds (event scheduled too far in future)")
```

Comment:
Do we need to do this on an error-handling path? It seems like it'd be useful to just clamp the timeout before we issue the call, no?

Reply:
I think its debatable as the bounds themselves are trawled up from kernel code. My approach was let the kernel say it's unhappy, and then suggest that exceeding these bounds is probably why. But I'm happy to put a check prior to calling kevent with more commenting that the bounds in kernel could change?

Reply:
I think that approach is better. It's less complex and it avoids concerns about kernel boundaries changing, which seems highly unlikely to me.

Reply:
OK. If you're setting a timer 2^29 ns into the future it was probably in error anyway, regardless of whether the kernel would accept it.

Reply:
It might be intentional, but the joys of clamping are that the penalty isn't high. If we really did sleep that long we'll wake up a few extra times, but we'll have been asleep for so long that those wakeup are essentially free.

Reply:
Just to be 100% clear by "clamp" you mean values greater than the kernel accepted maximum are limited to that value. And that, for example, specifying Int.max would not generate an error?

Reply:
Correct.